samza-core/src/main/java/org/apache/samza/storage/SideInputStorageManager.java

```diff
@@ -0,0 +1,372 @@
+              .filter(lastProcessedOffsets::containsKey)
+              .collect(Collectors.toMap(Function.identity(), lastProcessedOffsets::get));
+
+            try {
+              String checkpoint = OBJECT_MAPPER.writeValueAsString(offsets);
+              File offsetFile = new File(getStoreLocation(storeName), OFFSET_FILE);
+              FileUtil.writeWithChecksum(offsetFile, checkpoint);
+            } catch (Exception e) {
+              LOG.error("Encountered error while checkpointing to the file due to", e);
+              throw new SamzaException("Failed to checkpoint for side input store " + storeName, e);
+            }
+          });
+  }
+
+  /**
+   * Gets the side input SSP offsets for all stores from their local offset files.
+   *
+   * @return a {@link Map} of {@link SystemStreamPartition} to offset in the offset files.
+   */
+  @SuppressWarnings("unchecked")
+  private Map<SystemStreamPartition, String> getFileOffsets() {
+    LOG.info("Loading initial offsets from the file for side input stores.");
+    Map<SystemStreamPartition, String> fileOffsets = new HashMap<>();
+
+    stores.keySet().forEach(storeName -> {
+        LOG.debug("Reading local offsets for store {}", storeName);
+
+        File storeLocation = getStoreLocation(storeName);
+        if (!isValidSideInputStore(storeName)) {
+          try {
+            String checkpoint = StorageManagerUtil.readOffsetFile(storeLocation, OFFSET_FILE);
+            Map<SystemStreamPartition, String> offsets = OBJECT_MAPPER.readValue(checkpoint, Map.class);
+            fileOffsets.putAll(offsets);
+          } catch (Exception e) {
+            LOG.warn("Failed to load the checkpoints for store " + storeName, e);
+          }
+        }
+      });
+
+    return fileOffsets;
+  }
+
+  private File getStoreLocation(String storeName) {
+    return new File(storeBaseDir, (storeName + File.separator + taskName.toString()).replace(' ', '_'));
+  }
+
+  /**
+   * Initializes the starting offsets for the {@link SystemStreamPartition}s belonging to all the side input stores:
+   *   1. Gets the store offsets & filters out the offsets for stale stores
+   *   2. Gets the oldest SSP offsets from the source; e.g. Kafka
+   *   3. If locally checkpointed offset is available and is greater than the oldest available offset from source, pick it
+   *      Otherwise, fallback to oldest offset in the source.
+   *
+   * @return a {@link Map} of {@link SystemStreamPartition} to offset
+   */
+  private Map<SystemStreamPartition, String> getStartingOffsets() {
+    Map<SystemStreamPartition, String> fileOffsets = getFileOffsets();
+    Map<SystemStreamPartition, String> oldestOffsets = getOldestOffsets();
+
+    Map<SystemStreamPartition, String> startingOffsets = new HashMap<>();
+
+    sspsToStore.keySet().forEach(ssp -> {
+        String fileOffset = fileOffsets.get(ssp);
+        String oldestOffset = oldestOffsets.get(ssp);
```

Comment:
maybe log what the oldest offsets and fileOffsets are if it's not logged elsewhere?